From 55c8e8a67f5f960b7597032d3e290168641b531d Mon Sep 17 00:00:00 2001
From: Kirill Elagin <kirelagin@gmail.com>
Date: Sat, 15 Nov 2025 13:21:11 +0100
Subject: [PATCH] tests: Fix numpy type checking test

The testing code assumes that all numpy scalars have `__index__`,
however this is no longer true with recent numpy.

Backport the fix from upstream master by cherry-picking
64f5694cb2f9ae52ac6ffcad1eda4085e5cd5b7d.
---
 python/convert.c                              | 31 +++++++++++++++++++
 .../google/protobuf/internal/type_checkers.py | 13 ++++++--
 python/google/protobuf/pyext/message.cc       | 16 ++++++++--
 3 files changed, 56 insertions(+), 4 deletions(-)

diff --git a/python/convert.c b/python/convert.c
index 0c1174ba6..d4c17049e 100644
--- a/convert.c
+++ b/convert.c
@@ -226,6 +226,37 @@ bool PyUpb_IsNumpyNdarray(PyObject* obj, const upb_FieldDef* f) {
   return is_ndarray;
 }
 
+bool PyUpb_IsNumpyBoolScalar(PyObject* obj) {
+  PyObject* type_module_obj =
+      PyObject_GetAttrString((PyObject*)Py_TYPE(obj), "__module__");
+  bool is_numpy = !strcmp(PyUpb_GetStrData(type_module_obj), "numpy");
+  Py_DECREF(type_module_obj);
+  if (!is_numpy) {
+    return false;
+  }
+
+  PyObject* type_name_obj =
+      PyObject_GetAttrString((PyObject*)Py_TYPE(obj), "__name__");
+  bool is_bool = !strcmp(PyUpb_GetStrData(type_name_obj), "bool");
+  Py_DECREF(type_name_obj);
+  if (!is_bool) {
+    return false;
+  }
+  return true;
+}
+
+static bool PyUpb_GetBool(PyObject* obj, const upb_FieldDef* f, bool* val) {
+  if (!PyBool_Check(obj)) {
+    if (PyUpb_IsNumpyNdarray(obj, f)) return false;
+    if (PyUpb_IsNumpyBoolScalar(obj)) {
+      *val = PyObject_IsTrue(obj);
+      return !PyErr_Occurred();
+    }
+  }
+  *val = PyLong_AsLong(obj);
+  return !PyErr_Occurred();
+}
+
 bool PyUpb_PyToUpb(PyObject* obj, const upb_FieldDef* f, upb_MessageValue* val,
                    upb_Arena* arena) {
   switch (upb_FieldDef_CType(f)) {
@@ -248,9 +248,7 @@ bool PyUpb_PyToUpb(PyObject* obj, const upb_FieldDef* f, upb_MessageValue* val,
       val->double_val = PyFloat_AsDouble(obj);
       return !PyErr_Occurred();
     case kUpb_CType_Bool:
-      if (PyUpb_IsNumpyNdarray(obj, f)) return false;
-      val->bool_val = PyLong_AsLong(obj);
-      return !PyErr_Occurred();
+      return PyUpb_GetBool(obj, f, &val->bool_val);
     case kUpb_CType_Bytes: {
       char* ptr;
       Py_ssize_t size;

diff --git a/python/google/protobuf/internal/type_checkers.py b/python/google/protobuf/internal/type_checkers.py
index e152a43f8..a0e23a0ab 100755
--- a/google/protobuf/internal/type_checkers.py
+++ b/google/protobuf/internal/type_checkers.py
@@ -113,12 +113,21 @@ class BoolValueChecker(object):
   """Type checker used for bool fields."""
 
   def CheckValue(self, proposed_value):
-    if not hasattr(proposed_value, '__index__') or (
-        type(proposed_value).__module__ == 'numpy' and
+    if not hasattr(proposed_value, '__index__'):
+      # Under NumPy 2.3, numpy.bool does not have an __index__ method.
+      if (type(proposed_value).__module__ == 'numpy' and
+          type(proposed_value).__name__ == 'bool'):
+        return bool(proposed_value)
+      message = ('%.1024r has type %s, but expected one of: %s' %
+                 (proposed_value, type(proposed_value), (bool, int)))
+      raise TypeError(message)
+
+    if (type(proposed_value).__module__ == 'numpy' and
         type(proposed_value).__name__ == 'ndarray'):
       message = ('%.1024r has type %s, but expected one of: %s' %
                  (proposed_value, type(proposed_value), (bool, int)))
       raise TypeError(message)
+
     return bool(proposed_value)
 
   def DefaultValue(self):
diff --git a/python/google/protobuf/pyext/message.cc b/python/google/protobuf/pyext/message.cc
index 449a11d99..f2c6fb4b7 100644
--- a/google/protobuf/pyext/message.cc
+++ b/google/protobuf/pyext/message.cc
@@ -558,8 +558,20 @@ bool CheckAndGetFloat(PyObject* arg, float* value) {
 
 bool CheckAndGetBool(PyObject* arg, bool* value) {
   long long_value = PyLong_AsLong(arg);  // NOLINT
-  if (!strcmp(Py_TYPE(arg)->tp_name, "numpy.ndarray") ||
-      (long_value == -1 && PyErr_Occurred())) {
+  if (long_value == -1 && PyErr_Occurred()) {
+    // In NumPy 2.3, numpy.bool does not have an __index__ method and cannot
+    // be converted to a long using PyLong_AsLong.
+    if (!strcmp(Py_TYPE(arg)->tp_name, "numpy.bool")) {
+      PyErr_Clear();
+      int is_true = PyObject_IsTrue(arg);
+      if (is_true >= 0) {
+        *value = static_cast<bool>(is_true);
+        return true;
+      }
+    }
+    FormatTypeError(arg, "int, bool");
+    return false;
+  } else if (!strcmp(Py_TYPE(arg)->tp_name, "numpy.ndarray")) {
     FormatTypeError(arg, "int, bool");
     return false;
   }
-- 
2.51.0

